x-pack/libbeat/reader/parquet/parquet_benchmark_test.go

```diff
@@ -0,0 +1,307 @@
+			files:           []string{"cloudtrail.parquet"},
+			processParallel: true,
+			batchSize:       1,
+			invokeFn:        readParquetSingleRow,
+		},
+		{
+			desc:      "Construct a stream reader for a single file serially",
+			files:     []string{"cloudtrail.parquet"},
+			batchSize: 1,
+			invokeFn:  constructBufferedReader,
+		},
+		{
+			desc:            "Construct a stream reader for a single file parallelly",
+			files:           []string{"cloudtrail.parquet"},
+			processParallel: true,
+			batchSize:       1,
+			invokeFn:        constructBufferedReader,
+		},
+		{
+			desc:      "Construct a stream reader for multiple files serially",
+			files:     []string{"cloudtrail.parquet", "route53.parquet", "vpc_flow.gz.parquet"},
+			batchSize: 1,
+			invokeFn:  constructBufferedReader,
+		},
+		{
+			desc:            "Construct a stream reader for multiple files parallelly",
+			files:           []string{"cloudtrail.parquet", "route53.parquet", "vpc_flow.gz.parquet"},
+			processParallel: true,
+			batchSize:       1,
+			invokeFn:        constructBufferedReader,
+		},
+		{
+			desc:           "Construct a stream reader for multiple files using go routines",
+			files:          []string{"cloudtrail.parquet", "route53.parquet", "vpc_flow.gz.parquet"},
+			useGoRoutiunes: true,
+			batchSize:      1,
+			invokeFn:       constructBufferedReader,
+		},
+		{
+			desc:                       "Construct and read a single large parquet file in batches of 1000 serially",
+			constructAndReadLargeFiles: true,
+			largeFiles: []parquetFile{{
+				name: "large_file_1.parquet",
+				cols: 4,
+				rows: 100000,
+			}},
+			batchSize: 1000,
+			invokeFn:  readParquetFile,
+		},
+		{
+			desc:                       "Construct and read a single large parquet file in batches of 10000 serially",
+			constructAndReadLargeFiles: true,
+			largeFiles: []parquetFile{{
+				name: "large_file_2.parquet",
+				cols: 4,
+				rows: 100000,
+			}},
+			batchSize: 10000,
+			invokeFn:  readParquetFile,
+		},
+	}
+
+	for _, tc := range testCases {
+		b.Run(tc.desc, func(b *testing.B) {
+			cfg := &Config{
+				// we set ProcessParallel to true as this always has the best performance
+				ProcessParallel: true,
+				BatchSize:       tc.batchSize,
+			}
+
+			var files []string
+			if tc.constructAndReadLargeFiles {
+				tempDir := b.TempDir()
+				for _, f := range tc.largeFiles {
+					fName := tempDir + "/" + f.name
+					createRandomParquet(b, fName, f.cols, f.rows)
+					files = append(files, fName)
+				}
+			} else {
+				for _, f := range tc.files {
+					files = append(files, filepath.Join(testDataPath, f))
+
+				}
+			}
+
+			b.ResetTimer()
+			//nolint:errcheck // we do not care about handling errors from file.Seek()
+			switch {
+			case tc.processParallel:
+				b.RunParallel(func(pb *testing.PB) {
+					filePtrArr := openFiles(b, files)
+					for pb.Next() {
+						for _, f := range filePtrArr {
+							defer f.Close()
```

Comment:
Using a defer is a loop is usually not doing what is expected. I think that's the case here. This defer will execute when `RunParallel` returns. The same file would be `Closed()` multiple times, the same number of times that `pb.Next()` returns true.